### PR TITLE
 nixos/goldwarden: init, goldwarden: package GUI

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes

Things done:
- packaged GUI
- made a module which installs polkit files, pinentry
- tested ssh agent functionality
- tested browser functionality

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/goldwarden.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    enable = lib.mkEnableOption "Goldwarden";
+    package = lib.mkPackageOption pkgs "goldwarden" {};
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [
+      # for cli and polkit action
+      cfg.package
+      # binary exec's into pinentry which should match the DE
+      pkgs."pinentry-${config.programs.gnupg.agent.pinentryFlavor}"
```
Comment (wyndon) [inline nixos/modules/programs/goldwarden.nix:16]: This doesn't work right now (e.g. with gnome3), are we waiting for another PR ?

Comment (SuperSandro2000) [reply nixos/modules/programs/goldwarden.nix:16]: Yes, this requires #133542

Comment (MinerSebas) [inline nixos/modules/programs/goldwarden.nix:16]: Does this need to be in `environment.systemPackages`?
In my personal setup, adding a `path = [ pkgs."pinentry-${config.programs.gnupg.agent.pinentryFlavor}" ];` to the service also worked.

File: nixos/modules/programs/goldwarden.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    ];
+
+    # see https://github.com/quexten/goldwarden/blob/main/cmd/setup_linux.go
+    systemd.user.services.goldwarden = {
+      description = "Goldwarden daemon";
+      wantedBy = [ "graphical-session.target" ];
+      after = [ "graphical-session.target" ];
```
Comment (MinerSebas) [inline nixos/modules/programs/goldwarden.nix:23]: Inspired by the ssh-agent service https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/programs/ssh.nix#L327. This way this service is only run for non-system user accouts.
```suggestion
      after = [ "graphical-session.target" ];
      unitConfig.ConditionUser = "!@system";
```

File: nixos/modules/programs/goldwarden.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      serviceConfig.ExecStart = "${lib.getExe cfg.package} daemonize";
+    };
```
Comment (MinerSebas) [inline nixos/modules/programs/goldwarden.nix:25]: Necessary for ssh-agent functionality. 
```suggestion
    };
    
    environment.extraInit =
      ''
        if [ -z "$SSH_AUTH_SOCK" -a -n "$HOME" ]; then
          export SSH_AUTH_SOCK="$HOME/.goldwarden-ssh-agent.sock"
        fi
      '';
```
Inspired by the ssh-agent module. https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/programs/ssh.nix#L347-L352

File: nixos/modules/programs/goldwarden.nix (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/modules/programs/goldwarden.nix:16]: It is also used in the GUI and sometimes the client also starts the daemon on it's own and when it then is missing, things break.

File: nixos/modules/programs/goldwarden.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+let
+  cfg = config.programs.goldwarden;
+in
+{
+  options.programs.goldwarden = {
+    enable = lib.mkEnableOption "Goldwarden";
+    package = lib.mkPackageOption pkgs "goldwarden" {};
+  };
+
+  config = lib.mkIf cfg.enable {
```
Comment (MinerSebas) [inline nixos/modules/programs/goldwarden.nix:11]:
```suggestion
  config = lib.mkIf cfg.enable {
      assertions = [{ 
         assertion = !config.programs.ssh.startAgent;
         message = "Only one ssh-agent can be used at a time.";
      }];
```

Optionally also include another Option for not setting `SSH_AUTH_SOCK`, incase a User might want to use a different agent.

**Review by arthsmn — APPROVED** [review]

**Comment by Kasper24** [comment]
I'm running into an issue with the systemd service not having pinentry in PATH. Works fine when I run goldwarden daemonize manually though.
```
Error: exec: "pinentry": executable file not found in $PATH
```

**Review by arthsmn — APPROVED** [review]

**Review by arthsmn — APPROVED** [review]

File: nixos/modules/programs/goldwarden.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
+{ lib, config, pkgs, ... }:
+let
+  cfg = config.programs.goldwarden;
+in
+{
+  options.programs.goldwarden = {
+    enable = lib.mkEnableOption "Goldwarden";
+    package = lib.mkPackageOption pkgs "goldwarden" {};
+    useSshAgent = lib.mkEnableOption "Goldwarden's SSH Agent";
```
Comment (MinerSebas) [inline nixos/modules/programs/goldwarden.nix:9]: Personally, I would activate the ssh agent by default, but it's fine if it stays disabled by default.

**Review by MinerSebas — APPROVED** [review]

**Review by arthsmn — APPROVED** [review]
LGTM. If possible, can you add a simple test just to check if there are any breaking changes in the service?

**Comment by SuperSandro2000** [comment]
Goldwarden is sometimes a bit brittle and fails for no apparent reason. Also we would need a running vaultwarden and probably create some entries in it which is a bit more involved.

**Comment by quexten** [comment]
> Goldwarden is sometimes a bit brittle and fails for no apparent reason. Also we would need a running vaultwarden and probably create some entries in it which is a bit more involved.

**Comment by quexten** [comment]
> Goldwarden is sometimes a bit brittle and fails for no apparent reason. Also we would need a running vaultwarden and probably create some entries in it which is a bit more involved.

@SuperSandro2000 

On the GUI side, https://github.com/quexten/goldwarden/pull/191 should fix a lot of the issues regarding stability. The daemon still needs some work, if you do encounter further issues please try to report them though, so they can be fixed :)

**Pull request merged**